### DISCUSSION TITLE
How to make model more "sharp"?

### DISCUSSION THREAD

**Discussion by bingo-soft** [google/model-viewer#4794] [General]
By default my model looks very pale or vague. Just like this:

![model-viewer1](https://github.com/google/model-viewer/assets/48438172/7b673fc6-13d9-4c55-8dd5-a00c71209436)

For example, compare that to xeokit viewer:

![xeokit](https://github.com/google/model-viewer/assets/48438172/33bc612f-5139-4457-b042-7649ef44548a)

How can I get a look similar to xeokit with more contrast and more sharpness? Thank you in advance!

**Comment by elalish** [comment]
You are comparing to a non-physical rendering. The sharpness is due to rendering black lines on edges, which is not something glTF supports. Your colors are also probably not being exported correctly - maybe missing sRGB conversion or ignoring some other transform applied in your baseline renderer. Your materials also appear to be fully matte, which isn't very realistic - you'll get more realism by choosing appropriate roughness and metalness values.

**Comment by adamwolf** [comment]
I had a similar problem, actually, (not with colors, but with outlines).  I built a fragment shader that uses depth and normals to detect edges, and adds an outline.  (This approach can't be absolutely perfect, but I think it worked out well enough for what I need it for.) I rigged it into a custom pmndrs postprocessing effect, and then into a custom element so it works mostly like the built-in model-viewer-effects.  Unfortunately, I'm a little swamped this week with a deadline so I can't set it up for sharing right this moment.

(I think there's something I am missing with color-grading here, but I haven't looked into it yet.) Here's the same model with and without the new shader.

<img width="772" alt="Screenshot 2024-06-17 at 2 54 25 PM" src="https://github.com/google/model-viewer/assets/35681/92f6c88d-fd9d-4434-9201-e31b84ef0689">
<img width="790" alt="Screenshot 2024-06-17 at 2 54 20 PM" src="https://github.com/google/model-viewer/assets/35681/90f56fa3-ca1b-4801-af56-ede39f551a95">

**Reply by elalish** [reply]
That's amazing! Looking forward to taking a look when you're ready to share. In fact, this would make a wonderful addition to our examples if you'd be willing to make a PR. I've seen several requests from people wondering how to make their own custom post-processing effects.

**Reply by bingo-soft** [reply]
Agree. It looks really fantastic!

**Reply by george-martinec** [reply]
Please share code 🙏🏻 @adamwolf

**Reply by adamwolf** [reply]
I wrote an article about my work on this (https://www.atomwolf.org/posts/rendering-outlines-with-a-post-processing-shader/) a few months ago. I wrote that I wanted to package it up nicely in the "next steps" section, but then got snagged by other commitments. 

I'm able to spend some time on it over the holidays, and I'd love to wrap it up and have a nice example of a custom effect as a web component.
